### PR TITLE
 Introduce automatic pixel shifting 

### PR DESCRIPTION
**Pull request by jonas2515**
The panel in the touchbar is an OLED screen and pixels on it will suffer
degradation over time. Degradation effects are going to be worst on bright,
white pixels that remain at the same position forever. This applies to the
icons of our buttons.

To lower the possibilty of a burn-in here as much as we can, introduce some
subtle pixel shifting.

Here's how it works:

- Use the existing TIMEOUT_MS poll() timeout and (temporarily) allow smaller
  timeouts for smooth sub-pixel transitions to make the pixel shift completely
  invisible

- We have lots of space in the horizontal direction, so there we shift pixels
  by the whole amount that's necessary to ensure no pixel remains "always on"

- The background boxes around each button are also shifted in horizontal
  direction as otherwise the icons appear off-center horizontally, for this
  the boxes need to be reduced in width a bit so they don't overflow the panel

- In the vertical direction, icons appear off-center on the whole touchbar
  really quickly, so only do very limited movement here, that's not very
  effective but better than nothing

- Hitboxes for touch input remain the same and are not affected by pixel
  shifting

This starts a pixel shift by one full pixel in x and y direction every 10
seconds, transitioning slowly between the states on a sub-pixel level for
4 seconds. Once the pixel shift reaches the final left or right edge of
the shifting area, we'll leave it at that for 50 seconds to reduce stress
on pixels in the center of the shifting area a bit more.

https://github.com/WhatAmISupposedToPutHere/tiny-dfr/issues/9

### CODE REVIEW

File: src/pixel_shift.rs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+                self.x_progress += ANIMATION_INTERVAL_MS as f64 / ANIMATION_DURATION_MS as f64;
+            }
+            self.last_active = time_now;
+            pixels_changed = true;
+
+            if (self.x_progress % 1.0).abs() <= 0.01 || (self.x_progress % 1.0).abs() >= 0.99 {
+                self.x_progress = self.x_progress.round();
+                self.in_animation = false;
+                self.in_prolonged_timeout = false;
+                //println!("finished pixel shift, now {:?}", get_pixel_shift(self.x_progress, self.y_constant));
```
Comment (jonas2515) [inline src/pixel_shift.rs:96]: Not exactly sure what to do about those optional debug messages. I'm quite new to rust world, is there some kind of verbose debug function we could use for that?

**Comment by WhatAmISupposedToPutHere** [comment]
I have noticed that you completely disabled the button backgrounds. Is that a requirement for pixel shifting to look good?

**Comment by WhatAmISupposedToPutHere** [comment]
Also, how necessary it really is? Macos does not do it, and i was not able to find reports of burn-in being a problem. And here, the inactivity dimming is more aggressive than on macos.

**Comment by jonas2515** [comment]
> I have noticed that you completely disabled the button backgrounds. Is that a requirement for pixel shifting to look good?

Yeah, as mentioned on #9, that makes it a lot less noticable.

> Also, how necessary it really is? 

Right, this is somewhat optional. As mentioend on #9 I'd really like to have the touchbar always-on, as I think this is the thing that makes it actually useful as "dynamic" part of the keyboard. This is obviously my personal opinion and you're free to disagree.. I would have done this anyway for myself, so I thought I'd offer it upstream, too.

**Comment by WhatAmISupposedToPutHere** [comment]
> Yeah, as mentioned on #9, that makes it a lot less noticable.

Outlines are a hard requirement for me, so i will most likely end up introducing a config file and giving choices between outlines/no outlines and pixel shift/mandatory auto-dim. This will be merged after.

**Comment by jonas2515** [comment]
Ah, if you're open to that we could make always-on configurable and enable/disable outlines+pixel-shifting together with that?

Fwiw, I'd also really like to switch the default layer to the media keys instead of function keys, what do you think about that?

**Comment by WhatAmISupposedToPutHere** [comment]
> Fwiw, I'd also really like to switch the default layer to the media keys instead of function keys, what do you think about that?

Sure, i'll add that to the list of things to do.

**Comment by WhatAmISupposedToPutHere** [comment]
Sorry, i thought it would be faster to just change the things i wanted done in review myself. I have fixed the subpixel animations, but i am unable to reproduce the multiple-pixel-at-once y changes.

**Comment by jonas2515** [comment]
Hmm, subpixel animations seem fixed now, but y is still jumping around, and also seems like things can get stuck between two pixels on the x-axis, attached two debug logs of that.

[jumping-on-y-axis.log](https://github.com/WhatAmISupposedToPutHere/tiny-dfr/files/13308362/jumping-on-y-axis.log)
[stuck-between-11-and-12.log](https://github.com/WhatAmISupposedToPutHere/tiny-dfr/files/13308363/stuck-between-11-and-12.log)


This is the patch I applied to test



```patch
iff --git a/src/main.rs b/src/main.rs
index 738da0e..d1c5fee 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -44,7 +44,7 @@ const DFR_WIDTH: i32 = 2008;
 const DFR_HEIGHT: i32 = 64;
 const BUTTON_COLOR_INACTIVE: f64 = 0.200;
 const BUTTON_COLOR_ACTIVE: f64 = 0.400;
-const TIMEOUT_MS: i32 = 10 * 1000;
+const TIMEOUT_MS: i32 = 1 * 1000;
 
 #[derive(Deserialize)]
 #[serde(rename_all = "PascalCase")]
diff --git a/src/pixel_shift.rs b/src/pixel_shift.rs
index 634f722..c3af5b8 100644
--- a/src/pixel_shift.rs
+++ b/src/pixel_shift.rs
@@ -7,7 +7,7 @@ use crate::TIMEOUT_MS;
 const INTERVAL_MS: i32 = TIMEOUT_MS * 1; // should be a multiple of TIMEOUT_MS
 const PROLONGED_INTERVAL_MS: i32 = TIMEOUT_MS * 5; // should be a multiple of TIMEOUT_MS and more than INTERVAL_MS
 const ANIMATION_INTERVAL_MS: i32 = 200; // should be less than TIMEOUT_MS
-const ANIMATION_DURATION_MS: i32 = 4000; // should be a multiple of ANIMATION_INTERVAL_MS
+const ANIMATION_DURATION_MS: i32 = 400; // should be a multiple of ANIMATION_INTERVAL_MS
 
 // This is the total range on the x-axis that pixels will shift by over time, ie. they will shift by
 // PIXEL_SHIFT_WIDTH_PX / 2 to the right and to the left.
@@ -72,18 +72,22 @@ impl PixelShiftManager {
 
         match self.state {
             ShiftState::Normal => {
+            println!("starting subpixel shifting");
                 self.state = ShiftState::ShiftingSubpixel;
                 self.subpixel_progress = if self.direction == 1 { 0.0 } else { 1.0 };
             },
             ShiftState::ShiftingSubpixel => {
+                println!("shifting subpixel");
                 let shift_by = ANIMATION_INTERVAL_MS as f64 / ANIMATION_DURATION_MS as f64;
                 self.subpixel_progress += shift_by * self.direction as f64;
                 if self.subpixel_progress <= 0.01 || self.subpixel_progress >= 0.99 {
+                    println!("done with subpixel shift");
                     self.pixel_progress = (self.direction + self.pixel_progress as i64) as u64;
                     self.state = ShiftState::Normal;
                     self.subpixel_progress = 0.0;
                 }
                 if self.pixel_progress == 0 || self.pixel_progress >= PIXEL_SHIFT_WIDTH_PX {
+                    println!("done with cycle");
                     self.state = ShiftState::WaitingAtEnd;
                     self.direction = -self.direction;
                 }
@@ -98,6 +102,7 @@ impl PixelShiftManager {
     pub fn get(&self) -> (f64, f64) {
         let x_progress = self.pixel_progress as f64 + self.subpixel_progress;
         let y_progress = (x_progress + self.y_constant) % PIXEL_SHIFT_HEIGHT_PX as f64;
+        println!("x {}, y {}", x_progress - (PIXEL_SHIFT_WIDTH_PX / 2) as f64, y_progress - (PIXEL_SHIFT_HEIGHT_PX / 2) as f64);
         (x_progress - (PIXEL_SHIFT_WIDTH_PX / 2) as f64, y_progress - (PIXEL_SHIFT_HEIGHT_PX / 2) as f64)
     }
 }

**Comment by WhatAmISupposedToPutHere** [comment]
Ok, both should be fixed now

**Pull request merged**